url/src/lib.rs

```diff
@@ -145,29 +154,51 @@ url = { version = "2", features = ["debugger_visualizer"] }
 
 pub use form_urlencoded;
 
+// For forwards compatibility
+#[cfg(feature = "std")]
+extern crate std;
+
+#[macro_use]
+extern crate alloc;
+
+#[cfg(not(feature = "alloc"))]
+compile_error!("the `alloc` feature must be enabled");
```

Comment:
issue: we should just not have an alloc feature, and unconditionally use `alloc::Foo` and `core::Foo`.

Reply:
Happy to changes it, however I assumed a subset of features may be possible without alloc - and then introducing those will be another breaking change. Adding a feature flag isn't a lot of work

Reply:
Also, all other crates like idna and data-url already have the alloc feature

Reply:
If we wish to preplan for that I am happy to do so _if_ someone can sketch out a viable plan for it, otherwise it feels like a premature optimization.

Furthermore as I said before adding new default features is typically not considered breaking in the ecosystem, and I do not consider it breaking here.